src/Microsoft.OData.Client/Serialization/HttpWebRequestMessage.cs

```diff
@@ -5,207 +5,197 @@
 //---------------------------------------------------------------------
 
 namespace Microsoft.OData.Client
-{
+{ 
     using System;
     using System.Collections.Generic;
-    using System.Diagnostics;
-    using System.Diagnostics.CodeAnalysis;
-    using System.IO;
     using System.Globalization;
+    using System.IO;
+    using System.Linq;
     using System.Net;
-#if PORTABLELIB
-    using System.Reflection;
-    using System.Xml;
-#endif
+    using System.Net.Http;
+    using System.Net.Http.Headers;
+    using System.Threading.Tasks;
     using Microsoft.OData;
 
-    /// <summary> IODataRequestMessage interface implementation. </summary>
-    [SuppressMessage("Microsoft.Design", "CA1001:TypesThatOwnDisposableFieldsShouldBeDisposable", Justification = "Returning MemoryStream which doesn't require disposal")]
-    public class HttpWebRequestMessage : DataServiceClientRequestMessage
+    /// <summary>
+    /// HttpClient based implementation of DataServiceClientRequestMessage.
+    /// </summary>
+    public class HttpWebRequestMessage : DataServiceClientRequestMessage, IDisposable
     {
-        #region Private Fields
-
-#if PORTABLELIB
-        /// <summary> Running on silverlight </summary>
-        internal static bool IsRunningOnSilverlight = false;
-
-        /// <summary>Cached method info that will be used to call to set the userAgent property if its not null. </summary>
-        private static MethodInfo UserAgentMethodSetter;
-
-        /// <summary>Cached method info that will be used to call to set the ContentLength property if its not null. </summary>
-        private static MethodInfo ContentLengthMethodSetter;
-
-        /// <summary> Indicates the current platform the portablelib is running on can't set the user agent property </summary>
-        private static bool SettingUserAgentPropertyValidOnCurrentPlatform = true;
-
-        /// <summary> Indicates that the current platform doesn't support setting the header useragent property </summary>
-        private static bool SettingUserAgentOnHeadersCollectionValidOnCurrentPlatform = true;
-
-        /// <summary> Indicates the current platform the portablelib is running on can't set the user agent property </summary>
-        private static bool SettingContentLengthPropertyValidOnCurrentPlatform = true;
+        /// <summary>
+        /// HttpClient distinguishes "Content" headers from "Request" headers, so we
+        /// need to know which category a header belongs to.
+        /// </summary>
+        private static readonly IEnumerable<string> _contentHeaderNames = new[] {

```

Comment:
Consider using a case-insensitive HashSet later on the comparison can just be `_contentHeaderNames.Contains(headerName)`. Contains should also perform better than iterating the IEnumerable
```c#
new HashSet<string>(new []{...}, StringComparer.OrdinalIgnoreCase)
```

Reply:
solved